packages/pwa-kit-react-sdk/src/ssr/universal/hooks.js

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright (c) 2022, Salesforce, Inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import {useContext} from 'react'
+import {IsPrePassContext, ServerContext} from './contexts'
+
+/**
+ * Server context
+ * @typedef {Object} ServerContext
+ * @property {Object} req - Request object
+ * @property {Object} res - Response object
+ * @property {boolean} isServerSide
+ */
+
+/**
+ * Get the server context
+ * @returns {ServerContext} ServerContext object
+ *
+ * @example
+ * const {res, isServerSide} = useServerContext()
+ * if (isServerSide && query.error) { res.status(404) }
+ */
+export const useServerContext = () => {
+    const serverContext = useContext(ServerContext)
+    const isPrePass = useIsPrePass()
+
+    return {
+        ...serverContext,
+        isServerSide: Boolean(serverContext.req) && !isPrePass
```

Comment:
Another interesting thing to comment on, is that this hook kind of exposes the notion that we are doing or able to do a prepass on the server. Since the pre-pass logic for the with react query hoc is contained into one file, its probably not ideal to expose that information here. That being said, we are also relying on the context being undefined in the true render to have that conditional logic. 

Have you maybe thought about using the `res.locals` to get a variable before doing the prepass? 

For example you could potentially do something like this:

```
res.locals.isResReadOnly = true
await ssrPrepass(appJSX)
delete res.locals.isResReadOnly
```

Then in your hook you can access that value and run the callback if you plan on reverting how things currently work.

Reply:
I haven't thought of using `res.locals`.. will give that a try.

Reply:
Note: after further discussion, we decided to _not_ use `res.locals`. But clone the given `appJSX` instead, and make sure that the req/res is undefined for the prepass.